File: summary/ruby/21_local_search.rb
def local_search_algorithm(arr)
  n = arr.length
  best_solution = arr.dup
  best_cost = yield(best_solution)

  loop do
    current_solution = best_solution.dup
    improved = false

    (0...n).each do |i|
      (i+1...n).each do |j|
        current_solution[i], current_solution[j] = current_solution[j], current_solution[i]
        current_cost = yield(current_solution)
        if current_cost < best_cost
          best_solution = current_solution.dup
          best_cost = current_cost
          improved = true
        end
        current_solution[i], current_solution[j] = current_solution[j], current_solution[i]
      end
    end

    break unless improved
  end

  best_solution
end

# Example usage:
input_array = gets.chomp.split.map(&:to_i)
best_array = local_search_algorithm(input_array) { |arr| arr.reduce(0) { |sum, num| sum + num ** 2 } }
puts "Optimized Array: #{best_array}"

File: summary/ts/21_local_search.ts
function localSearchAlgorithm(arr: number[], costFunction: (arr: number[]) => number): number[] {
    const n: number = arr.length;
    let bestSolution: number[] = arr.slice();
    let bestCost: number = costFunction(bestSolution);

    while (true) {
        let currentSolution: number[] = bestSolution.slice();
        let improved: boolean = false;

        for (let i = 0; i < n; i++) {
            for (let j = i + 1; j < n; j++) {
                [currentSolution[i], currentSolution[j]] = [currentSolution[j], currentSolution[i]];
                const currentCost: number = costFunction(currentSolution);
                if (currentCost < bestCost) {
                    bestSolution = currentSolution.slice();
                    bestCost = currentCost;
                    improved = true;
                }
                [currentSolution[i], currentSolution[j]] = [currentSolution[j], currentSolution[i]];
            }
        }

        if (!improved) {
            break;
        }
    }

    return bestSolution;
}

// Example usage:
const inputArray: number[] = [3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5];
const bestArray: number[] = localSearchAlgorithm(inputArray, (arr) => arr.reduce((sum, num) => sum + num ** 2, 0));
console.log(`Optimized Array: ${bestArray}`);
